pyparsing/core.py

```diff
@@ -4511,44 +4515,40 @@ def recurse(self) -> List[ParserElement]:
         return [self.expr] if self.expr is not None else []
 
     def parseImpl(self, instring, loc, doActions=True):
-        if self.expr is not None:
-            try:
-                return self.expr._parse(instring, loc, doActions, callPreParse=False)
-            except ParseBaseException as pbe:
-                if not isinstance(self, Forward) or self.customName is not None:
-                    pbe.msg = self.errmsg
-                raise
-        else:
+        if self.expr is None:
             raise ParseException(instring, loc, "No expression defined", self)
 
+        try:
+            return self.expr._parse(instring, loc, doActions, callPreParse=False)
+        except ParseBaseException as pbe:
+            if not isinstance(self, Forward) or self.customName is not None:
+                pbe.msg = self.errmsg
+            raise
+
     def leave_whitespace(self, recursive: bool = True) -> ParserElement:
         super().leave_whitespace(recursive)
 
-        if recursive:
-            if self.expr is not None:
-                self.expr = self.expr.copy()
-                self.expr.leave_whitespace(recursive)
+        if recursive and self.expr is not None:
+            self.expr = self.expr.copy()
+            self.expr.leave_whitespace(recursive)
+
         return self
 
     def ignore_whitespace(self, recursive: bool = True) -> ParserElement:
         super().ignore_whitespace(recursive)
 
-        if recursive:
-            if self.expr is not None:
-                self.expr = self.expr.copy()
-                self.expr.ignore_whitespace(recursive)
+        if recursive and self.expr is not None:
+            self.expr = self.expr.copy()
+            self.expr.ignore_whitespace(recursive)
+
         return self
 
     def ignore(self, other) -> ParserElement:
-        if isinstance(other, Suppress):
```

Comment:
Why not `if isinstance(other, Suppress) and other not in self.ignoreExprs:` ? Negative logic looks weird to me here.